### PR TITLE
Deny requests without metadata header to avoid SSRF

### PR DESCRIPTION
**Pull request by hellokangning**
**Reason for Change**:

This PR denies the requests without `metadata` header to avoid SSRF, per [Get a token using HTTP](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http).

**Issue Fixed**:

[#514](https://github.com/Azure/aad-pod-identity/issues/514)

### CODE REVIEW

File: pkg/nmi/server/server.go [hunk 1]
```diff
@@ -308,6 +313,23 @@ func (s *Server) getTokenForExceptedPod(rqClientID, rqResource string) ([]byte,
...
+	metadata := parseMetadata(r)
+	if metadata != "true" {
+		klog.Error("metadata header is not specified")
+		metadataResp := MetadataResponse{
+			Error:            "invalid_request",
+			ErrorDescription: "Required metadata header not specified",
+		}
+		response, err := json.Marshal(metadataResp)
+		if err != nil {
+			klog.Errorf("failed to marshal metadata response, %+v", err)
```
Comment (hellokangning) [inline pkg/nmi/server/server.go]: This should never happen

**Comment by hellokangning** [comment]
> Thank you for the PR. Looked at the changes and they look good to me. Would you be able to add a new unit test to the PR? If I remember correctly, you should be able to create a new fake server with the app handler that you want to test, send a GET request and see the handler return the error. You can create a new mux server and then pass that mux to `httptest.NewServer` ("net/http/httptest"). This should help you test the handler with a fake server.

Thanks Anish, Updated with UT

File: docs/readmes/README.featureflags.md [hunk 2]
```diff
@@ -48,3 +48,11 @@ their use cases.
...
 Aad-pod-identity has a new flag `immutable-user-msis` which can be used to prevent deletion of specified identities from VM/VMSS.
 The list is comma separated. Example: 00000000-0000-0000-0000-000000000000,11111111-1111-1111-1111-111111111111
+
+## Metadata header required flag
+
+When you query the Instance Metadata Service, you must provide the header `Metadata: true` to ensure the request was not unintentionally redirected. Your can see examples of this header in [the Azure documentation](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service#using-headers).
+
+This is critical especially when you [acquire an access token](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http) as a mitigation againt Server Side Request Forgery (SSRF) attack. 
+
+The `metadataHeaderRequired` flag for NMI will block all requests without Metadata header and return an HTTP 400 response. This flag is disabled by default for compability, but recommended for usrs to enable this feature.
```
Comment (aramase) [inline docs/readmes/README.featureflags.md]: nit: s/usr/users

Comment (aramase) [inline docs/readmes/README.featureflags.md]: can you also add the flag used to enable outside of helm? 

Something like `Add the following arg --metadata-header-required=true to NMI daemonset`

File: pkg/nmi/server/server_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+func TestMsiHandler_NoMetadataHeader(t *testing.T) {
+	setup()
+	defer teardown()
+
+	s := &Server{
+		MetadataHeaderRequired: true,
+	}
+	mux.Handle(tokenPath, appHandler(s.msiHandler))
+
+	req, err := http.NewRequest("GET", tokenPath, nil)
```
Comment (aramase) [inline pkg/nmi/server/server_test.go]: can we use `http.MethodGet` instead of `GET`?

File: docs/readmes/README.featureflags.md (hunk 2 above)
Comment (hellokangning) [reply docs/readmes/README.featureflags.md]: Done, but set is to `false` for compability

File: go.mod [hunk 4]
```diff
@@ -30,6 +30,7 @@ require (
 	go.opencensus.io v0.22.0
 	golang.org/x/sync v0.0.0-20190227155943-e225da77a7e6
 	golang.org/x/time v0.0.0-20180412165947-fbb02b2291d2 // indirect
+	google.golang.org/api v0.4.0
```
Comment (kkmsft) [inline go.mod]: Any idea how this library was required ?

File: deploy/infra/deployment.yaml [hunk 5]
```diff
@@ -82,6 +82,7 @@ spec:
         args:
           - "--host-ip=$(HOST_IP)"
           - "--node=$(NODE_NAME)"
+          - --metadata-header-required=false
```
Comment (aramase) [inline deploy/infra/deployment.yaml]: We can't update the deployment yaml until a new release is made. So can you please revert this change?

Comment (hellokangning) [reply deploy/infra/deployment.yaml]: Sorry for misunderstanding, reverted

File: go.mod (hunk 4 above)
Comment (hellokangning) [reply go.mod]: I tried to use `testing` in this module and switched to [official testing](https://golang.org/pkg/testing/). Remove this

File: docs/readmes/README.featureflags.md [hunk 6]
```diff
@@ -48,3 +48,11 @@ their use cases.
...
 Aad-pod-identity has a new flag `immutable-user-msis` which can be used to prevent deletion of specified identities from VM/VMSS.
 The list is comma separated. Example: 00000000-0000-0000-0000-000000000000,11111111-1111-1111-1111-111111111111
+
+## Metadata header required flag
+
+When you query the Instance Metadata Service, you must provide the header `Metadata: true` to ensure the request was not unintentionally redirected. Your can see examples of this header in [the Azure documentation](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service#using-headers).
+
+This is critical especially when you [acquire an access token](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http) as a mitigation againt Server Side Request Forgery (SSRF) attack. 
+
+The `metadataHeaderRequired` flag for NMI will block all requests without Metadata header and return an HTTP 400 response. This flag is disabled by default for compability, but recommended for users to enable this feature.
```
Comment (aramase) [inline docs/readmes/README.featureflags.md]: s/compability/compatibility

File: docs/readmes/README.featureflags.md [hunk 7]
```diff
@@ -48,3 +48,11 @@ their use cases.
 
 Aad-pod-identity has a new flag `immutable-user-msis` which can be used to prevent deletion of specified identities from VM/VMSS.
 The list is comma separated. Example: 00000000-0000-0000-0000-000000000000,11111111-1111-1111-1111-111111111111
+
+## Metadata header required flag
+
+When you query the Instance Metadata Service, you must provide the header `Metadata: true` to ensure the request was not unintentionally redirected. Your can see examples of this header in [the Azure documentation](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service#using-headers).
```
Comment (aramase) [inline docs/readmes/README.featureflags.md]: s/Your/You

File: docs/readmes/README.featureflags.md [hunk 8]
```diff
@@ -48,3 +48,11 @@ their use cases.
 
 Aad-pod-identity has a new flag `immutable-user-msis` which can be used to prevent deletion of specified identities from VM/VMSS.
 The list is comma separated. Example: 00000000-0000-0000-0000-000000000000,11111111-1111-1111-1111-111111111111
+
+## Metadata header required flag
```
Comment (aramase) [inline docs/readmes/README.featureflags.md]: Also, after this can you add `> Available from 1.6 release` similar to the rest of the features in the readme

Comment (hellokangning) [reply docs/readmes/README.featureflags.md]: Done, thanks for your carefulness and patience

File: pkg/nmi/server/server.go [hunk 9]
```diff
@@ -372,7 +384,33 @@ func (s *Server) msiHandler(w http.ResponseWriter, r *http.Request) (ns string)
...
+	}
+	response, err := json.Marshal(metadataResp)
+	if err != nil {
+		klog.Errorf("failed to marshal metadata response, %+v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("Server", "Microsoft-IIS/10.0")
```
Comment (kkmsft) [inline pkg/nmi/server/server.go]: Could we please not return IIS header, instead leave the default header returned. In the past this has helped us determine if there tokens came from IMDS or aad-pod-identity.

**Comment by hellokangning** [comment]
> 
> 
> @hellokangning - A small request to change the return header to default. Other wise /lgtm. Thank you !

Done
